### ISSUE TITLE
[SSR] Recoil doesn't work server side with Next.js 13

### ISSUE THREAD

**Issue by leonbe02** [facebookexperimental/Recoil#2082]
## The Gist

When using Recoil + Next.js 13, I'm encountering an error while pre-rendering the page server side. The app works as expected client side though so no issue there, just when Recoil is used during server side pre-render.

## The Error
```
TypeError: batcher is not a function
    at MutableSnapshot.batchUpdates [as _batch] (webpack-internal:///(sc_client)/./node_modules/recoil/cjs/index.js:3447:3)
    at eval (webpack-internal:///(sc_client)/./node_modules/recoil/cjs/index.js:3981:12)
    at eval (webpack-internal:///(sc_client)/./app/RecoilProvider.tsx:17:9)
    at Snapshot.eval [as map] (webpack-internal:///(sc_client)/./node_modules/recoil/cjs/index.js:3749:7)
    at freshSnapshot (webpack-internal:///(sc_client)/./node_modules/recoil/cjs/index.js:3939:45)
    at initialStoreState (webpack-internal:///(sc_client)/./node_modules/recoil/cjs/index.js:4397:20)
    at eval (webpack-internal:///(sc_client)/./node_modules/recoil/cjs/index.js:4548:187)
    at useRefInitOnce (webpack-internal:///(sc_client)/./node_modules/recoil/cjs/index.js:4078:19)
    at RecoilRoot_INTERNAL (webpack-internal:///(sc_client)/./node_modules/recoil/cjs/index.js:4548:19)
    at renderWithHooks (node_modules/next/dist/compiled/react-dom/cjs/react-dom-server.browser.development.js:7621:16)
```

## The Environment
- next: 13.0.0
- react: 18.2.0
- react-dom: 18.2.0
- recoil: 0.7.6

## The Code
```tsx
// app/layout.tsx

'use client';

import RecoilProvider from './RecoilProvider';

export default function RootLayout(
  { children }: { children: JSX.Element }
) {
  return (
    <html lang="en-us">
      <head>
        <title>Next 13 + Recoil Example</title>
      </head>
      <body>
        <RecoilProvider locale="en-us">{children}</RecoilProvider>
      </body>
    </html>
  );
}
```

```tsx
// app/atoms/i18n.tsx

'use client';

import { atom } from 'recoil';

export const locale = atom({
  key: 'locale',
  default: 'en-us',
});
```

```tsx
// app/page.tsx

'use client';

import { useRecoilValue } from 'recoil';
import * as i18n from './atoms/i18n';

export default function Home() {
  const locale = useRecoilValue(i18n.locale);

  return (
    <main>
      <p>The locale: {locale}</p>
    </main>
  );
}
```

```tsx
// app/RecoilProvider.tsx

'use client';

import { useCallback } from 'react';
import { RecoilRoot, SetRecoilState } from 'recoil';
import * as i18n from './atoms/i18n';

export default function RecoilProvider({
  locale,
  children,
}: {
  locale: string,
  children: JSX.Element,
}) {
  const initializeState = useCallback(({ set }: { set: SetRecoilState }) => {
    set(i18n.locale, locale);
  }, [locale]);

  return (
    <RecoilRoot initializeState={initializeState}>
      {children}
    </RecoilRoot>
  )
}
```

**Comment by hu-tu-yi-shi** [comment]
same here

**Comment by Soumajit2004** [comment]
Same here

**Comment by tpAtalas** [comment]
This is mysterious. Mine works fine and I have exactly same setting. Will update if I find something.

**Comment by 398noe** [comment]
same for me :(

**Comment by leonbe02** [comment]
Upon further investigation, the issue appears to be specific to using the `initializeState` property on `RecoilRoot`. If I remove the function that initializes the atom, the page renders server side just fine. When I pass in a function and attempt to initialize the state, I get the `batcher is not a function` error.

**Comment by leonbe02** [comment]
I've pinpointed the issue I believe to the usage of `unstable_batchedUpdates` from `react-dom`: https://github.com/facebookexperimental/Recoil/blob/main/packages/recoil/core/Recoil_Batching.js#L14

The version of `react-dom` that Next 13 is using is `18.3.0-next-d925a8d0b-20221024` which doesn't appear to have the `unstable_batchedUpdates` function anymore. Are there plans to remove usage of that function in favor of something else?

**Comment by leonbe02** [comment]
@drarmstr Any estimate on when this change will be released in a new NPM version?

**Comment by kilofeq** [comment]
Hi, @leonbe02. Did you find any workaround? I'm struggling with migrating to Next.js 13 :/

**Comment by devcaeg** [comment]
Any news about not being able to initialize Recoil with data obtained in SSR in Next.js 13?

**Comment by andreisoare** [comment]
@drarmstr would you be able to help make a release that includes this fix? Still blocked by this. Or is there someone else who can help? Thank you!

**Comment by rootical** [comment]
Another vote for a release...

**Comment by leonbe02** [comment]
Has this project died or something? There hasn't been a new version released since October of last year. I know Facebook had layoffs but I didn't realize they laid off the entire team behind Recoil...

**Comment by jephjohnson** [comment]
Any updates on this? Same Issues

**Comment by Mng12345** [comment]
Same for me. Does it mean that the component has a head line with `"use client"` would not be rendered in the server side?

**Comment by RizqiSyahrendra** [comment]
for this issue, I ended up manually waiting the client ready before rendering from recoil state, I did this because the content I was rendering is not required to be SSR

**Comment by khashayarghajar** [comment]
any news ?

**Comment by DjibrilM** [comment]
> Just finished migrating one of my projects to 13. This is what I found:
> 
> * All work fine as long as components using Recoil hooks have the directive 'use client'. That includes all the child components.
> * Using the 'use client' directive in layout ONLY will fail to load as mentioned above.
> * Now fetching with atomEffect works without turning off SSR through the dynamic import module, but will fetch twice... except the one pre-fetching with useRecoilCallback.
> * Does not need to use the 'use client' directive to atoms/selectors/custom hook using useRecoilCallback if they are in separate files. It is fine if the client directive is properly used within the component.
> 
> All work is the same as Next12 on my side except for some non-Recoil-related errors, such as some 'Modules are not found.' This is a known issue, I believe.
> 
> Next13's app directory is far from ready to use, and I will skip using it until it becomes stable.

then why should we use Nextjs if we shall end up turning everything into a client component?

**Comment by drarmstr** [comment]
Fixed with #2086 and released with `0.7.7`

**Comment by elmcapp** [comment]
This is not an issue. The way that Next.js works is server-side first. You can't just wrap the app in the ```<RecoilRoot>``` You do have to have use the follow ```"use client"``` however you have to put it in a wrapper component if you are using the app directory. According to Next.js state including state management must be handled on the client as there is no way to have state on the server side. Below is a working example that works for me

1. Create a wrapper file called RecoilRootWrapper.js
````
"use client";

import React from "react";
import { RecoilRoot } from "recoil";

function RecoilRootWrapper({ children }) {
  return <RecoilRoot >{children}</RecoilRoot>;
}

export default RecoilRootWrapper;

````

2. In the app directory edit the layout.tsx or layout.js file
```
- src/
  - app/
    - pages.tsx
    - layout.tsx
 ```
 layout.tsx file should look like this:
 ````
import RecoilRootWrapper from "@/wrappers/RecoilRootWrapper";
import "./globals.css";
import { Inter } from "next/font/google";

const inter = Inter({ subsets: ["latin"] });

export const metadata = {
  title: "Create Next App",
  description: "Generated by create next app",
};

export default function RootLayout({
  children,
}: {
  children: React.ReactNode;
}) {
  return (
    <html lang="en">
      <body className={inter.className}>
        <RecoilRootWrapper>{children}</RecoilRootWrapper>
      </body>
    </html>
  );
}

````

By doing this we put the Recoil provider on the client (browser) and anything that is not marked with "use client" will be rendered as server side. You can not add "use client" to the layout.tsx or layout.js, page.tsx or page.js files

**Comment by MattSteedman** [comment]
> This is not an issue. The way that Next.js works is server-side first. You can't just wrap the app in the `<RecoilRoot>` You do have to have use the follow `"use client"` however you have to put it in a wrapper component if you are using the app directory. According to Next.js state including state management must be handled on the client as there is no way to have state on the server side. Below is a working example that works for me
> 
> 1. Create a wrapper file called RecoilRootWrapper.js
> 
> ```
> "use client";
> 
> import React from "react";
> import { RecoilRoot } from "recoil";
> 
> function RecoilRootWrapper({ children }) {
>   return <RecoilRoot >{children}</RecoilRoot>;
> }
> 
> export default RecoilRootWrapper;
> ```
> 
> 2. In the app directory edit the layout.tsx or layout.js file
> 
> ```
> - src/
>   - app/
>     - pages.tsx
>     - layout.tsx
> ```
> 
> layout.tsx file should look like this:
> 
> ```
> import RecoilRootWrapper from "@/wrappers/RecoilRootWrapper";
> import "./globals.css";
> import { Inter } from "next/font/google";
> 
> const inter = Inter({ subsets: ["latin"] });
> 
> export const metadata = {
>  title: "Create Next App",
>  description: "Generated by create next app",
> };
> 
> export default function RootLayout({
>  children,
> }: {
>  children: React.ReactNode;
> }) {
>  return (
>    <html lang="en">
>      <body className={inter.className}>
>        <RecoilRootWrapper>{children}</RecoilRootWrapper>
>      </body>
>    </html>
>  );
> }
> ```
> 
> By doing this we put the Recoil provider on the client (browser) and anything that is not marked with "use client" will be rendered as server side. You can not add "use client" to the layout.tsx or layout.js, page.tsx or page.js files


Although this might work as a hack there are some files or compomponents you want server-side so wont this wrapper essentially put all those {children} in the clientside in your layout.tsx file

**Comment by elmcapp** [comment]
@MattSteedman Nope even if you put it as a wrapper any child components will still be server unless you use "use client" in each component. This is not a hack or workaround. Recoil is state management tools and anything that use state must be rendered on the client.

**Comment by Yedidya10** [comment]
Hi, There is another issue happening because of SSR

If I want to use the effects property, even though I specify 'use client' in the component I wrap in `<RecoilRoot>`, I still get an error:

> Unhandled Runtime Error
Error: async/await is not yet supported in Client Components, only Server Components. This error is often caused by accidentally adding `'use client'` to a module that was originally written for the server.

**Is there a solution for this, or is there a PR about it?**

```
import { AtomEffect, atom } from 'recoil'

export type ThemeMode = 'light' | 'dark'

const localStorageEffect =
  (key: string): AtomEffect<ThemeMode> =>
  ({ setSelf, onSet }) => {
    // Retrieve the value stored at the specified key
    const stored = localStorage.getItem(key)
    // Check if the value exists and is light or dark
    if (stored === 'dark' || stored === 'light') {
      // If the value is valid, the call the provided function setSelf which initializes the atom value
      setSelf(stored)
    }
    // Creates the callback triggered when the atom is changed
    onSet((value, _, isReset) => {
      if (isReset) {
        // If atom has been reset then remove it from local storage
        localStorage.removeItem(key)
      } else {
        // If value has changed then store the value in local storage
        localStorage.setItem(key, value || _) // the || is a fail-safe if for any reason value is null the value will revert to default
      }
    })
  }

const themeModeState = atom<ThemeMode>({
  key: 'themeModeState',
  default: 'light',
  effects: [localStorageEffect('themeMode')],
})

export default themeModeState
```

**Comment by bennidhamma** [comment]
Is there any way to share recoil state across multiple recoil roots? This suggests no:

> Multiple RecoilRoot's may co-exist and represent independent providers/stores of atom state; atoms will have distinct values within each root.

https://recoiljs.org/docs/api-reference/core/RecoilRoot/

Does this mean recoil state can only apply to a single client subtree of a next app? My use case is that I have a SSR app, with different tree nodes rendered on client using the 'use client' directive. 

```
Server Component
 |     |     |
 A     B     C
```

Where child components A and C are client components. Is it possible to share recoil state between A and C? it seems like that could theoretically work - they are both in the same browser process after all.... 

(this seems like a pretty common use case for state - updating a navigation sidebar if a item is consumed / interacted, or a shopping cart count, etc....)
